File: rust/day5_part2_2017.rs

use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").expect("Unable to read file");
    let mut jumps: Vec<i32> = input.lines().map(|x| x.parse().unwrap()).collect();
    
    let mut steps = 0;
    let mut current_index = 0;
    
    while current_index >= 0 && current_index < jumps.len() as i32 {
        let jump = jumps[current_index as usize];
        
        if jump >= 3 {
            jumps[current_index as usize] -= 1;
        } else {
            jumps[current_index as usize] += 1;
        }
        
        current_index += jump;
        steps += 1;
    }
    
    println!("{}", steps);
}

File: training_data/reflection-tuning/day5_part2_2017.clj
(ns maze-solver
  (:require [clojure.string :as str]))

(defn solve [jumps part]
  (loop [pos 0
         steps 0
         jumps (transient jumps)]
    (if (or (< pos 0) (>= pos (count jumps)))
      steps
      (let [jump (jumps pos)
            new-jump (if (= part 2)
                       (if (>= jump 3) (dec jump) (inc jump))
                       (inc jump))]
        (recur (+ pos jump)
               (inc steps)
               (assoc! jumps pos new-jump))))))

(defn -main []
  (let [jumps (mapv parse-long (str/split-lines (slurp "input.txt")))]
    (println "Part 1:" (solve jumps 1))
    (println "Part 2:" (solve jumps 2))))

(-main)
